compiler/pipes/final-check.cpp

```diff
@@ -42,11 +42,37 @@ void check_class_immutableness(ClassPtr klass) {
 void process_job_worker_class(ClassPtr klass) {
   auto request_interface = G->get_class("KphpJobWorkerRequest");
   auto response_interface = G->get_class("KphpJobWorkerResponse");
-  if ((request_interface && request_interface->is_parent_of(klass)) ||
-      (response_interface && response_interface->is_parent_of(klass))) {
+  auto shared_memory_piece_interface = G->get_class("KphpJobWorkerSharedMemoryPiece");
+  kphp_assert(request_interface);
+  kphp_assert(response_interface);
+  kphp_assert(shared_memory_piece_interface);
+
+  bool implements_request = request_interface->is_parent_of(klass);
+  bool implements_response = response_interface->is_parent_of(klass);
+  bool implements_shared_memory_piece = shared_memory_piece_interface->is_parent_of(klass);
+
+  if (implements_request || implements_response || implements_shared_memory_piece) {
     klass->deeply_require_instance_cache_visitor();
     klass->deeply_require_virtual_builtin_functions();
   }
+  if (implements_shared_memory_piece) {
+    kphp_error(klass->is_immutable, fmt_format("Class {} must be immutable (@kphp-immutable-class) as it implements KphpJobWorkerSharedMemoryPiece", klass->name));
+  }
+  if (implements_request) {
+    size_t shared_memory_piece_members_count = 0;
+    for (ClassPtr ancestor : klass->get_all_ancestors()) {
+      ancestor->members.for_each([&](const ClassMemberInstanceField &field) {
+        for (ClassPtr field_class : field.get_inferred_type()->class_types()) {
+          if (shared_memory_piece_interface->is_parent_of(field_class)) {
+            ++shared_memory_piece_members_count;
+            G->add_job_worker_shared_memory_piece(klass->name, &field);
```

Comment:
usage std::map in `add_job_worker_shared_memory_piece()` looks like a race condition :(
May be put this field directly into class ClassData (ClassPtr) object? It will be a bit simpler and without races :)

Reply:
fixed